tianshou/env/venvs.py

```diff
@@ -202,16 +202,25 @@ def reset(
             self.workers[i].send(None, **kwargs)
         ret_list = [self.workers[i].recv() for i in id]
 
-        if "return_info" in kwargs and kwargs["return_info"]:
+        has_infos = isinstance(ret_list[0], (tuple, list)) and len(
+            ret_list[0]
+        ) == 2 and isinstance(ret_list[0][1], dict)
+        if has_infos:
```

Comment:
Can we use `self.has_infos` to reduce checking overhead?
```python
if self.has_info is None:
  self.has_info = self.check_has_info(ret_list[0])

if self.has_info:
  ...
else:
  ...
```

Reply:
implemented in https://github.com/thu-ml/tianshou/pull/613/commits/662a68a05363db391b2a50717010f41ae10d7f1f. (although I did `if hasattr(self, reset_returns_info)`, as mypy complains about typing in a bunch of places with `self.has_info` is None). Although a warning that with this change, users may experience a surprise if they call `reset` with `return_info` set to `True` and then `False`.

Reply:
Have you ever tested with a full training pipeline?

Reply:
> Have you ever tested with a full training pipeline?
You mean run it with one of the scripts in the examples folder? Haven't done that yet but can do.

I'm not sure if the second approach would work in all cases, since there could be some environments that don't return info along with obs right now. It would also be jumping ahead of the planned Gym API changes. If the 3rd option also sound good with you, I can implement that.

Reply:
I mean if there's no info, we attach an empty dict.

Reply:
> If the 3rd option also sound good with you, I can implement that.

I'm ok with it.

Reply:
added support to collector for option 3 in https://github.com/thu-ml/tianshou/pull/613/commits/75ecd180f7f38076ad10e36301b367b2cf63268f

Also I ran `python examples/box2d/lunarlander_dqn.py` and it works fine